mindmeld/resource_loader.py

```diff
@@ -40,7 +40,176 @@
+
+    def __len__(self):
+        return len(self.row_ids)
+
+    def __bool__(self):
+        return bool(len(self))
+
+    def processed_queries(self):
+        return ProcessedQueryList.Iterator(self)
+
+    def raw_queries(self):
+        return ProcessedQueryList.RawQueryIterator(self)
+
+    def queries(self):
+        return ProcessedQueryList.QueryIterator(self)
+
+    def entities(self):
+        return ProcessedQueryList.EntitiesIterator(self)
+
+    def domains(self):
+        return ProcessedQueryList.DomainIterator(self)
+
+    def intents(self):
+        return ProcessedQueryList.IntentIterator(self)
+
+    @staticmethod
+    def from_in_memory_list(queries):
+        '''
+        Creates a ProcessedQueryList wrapper around an
+        in-memory list of ProcessedQuery objects
+
+        Args:
+            queries (list(ProcessedQuery)): queries to wrap
+        Returns:
+            ProcessedQueryList object
+        '''
+        return ProcessedQueryList(
+            cache=ProcessedQueryList.MemoryCache(queries),
+            row_ids=list(range(len(queries)))
+        )
+
+    class Iterator:
+        def __init__(self, source, cached=False):
+            self.source = source
+            self.row_ids = source.row_ids
+            self.result_cache = [] if not cached else [None] * len(source)
+            self.iter_idx = -1
+
+        def __iter__(self):
+            self.iter_idx = -1
+            return self
+
+        def __next__(self):
+            self.iter_idx += 1
+            if self.iter_idx == len(self):
+                raise StopIteration
+            if self.result_cache and self.result_cache[self.iter_idx] is not None:
+                return self.result_cache[self.iter_idx]
+            result = self[self.iter_idx]
+            if self.result_cache:
+                self.result_cache[self.iter_idx] = result
+            return result
+
+        def __len__(self):
+            return len(self.source)
+
+        def __getitem__(self, key):
+            '''
+            Override this function for getting other types of cached data.
+            '''
+            return self.source.cache.get(self.row_ids[key])
+
+        def reorder(self, indices):
+            self.row_ids = [self.row_ids[i] for i in indices]
+            if self.result_cache:
+                self.result_cache = [self.result_cache[i] for i in indices]
+
+    class RawQueryIterator(Iterator):
+        def __getitem__(self, key):
+            return self.source.cache.get_raw_query(self.row_ids[key])
+
+    class QueryIterator(Iterator):
+        def __getitem__(self, key):
+            return self.source.cache.get_query(self.row_ids[key])
+
+    class EntitiesIterator(Iterator):
+        def __getitem__(self, key):
+            return self.source.cache.get_entities(self.row_ids[key])
+
+    class DomainIterator(Iterator):
+        def __init__(self, source):
+            cached = not isinstance(source.cache, ProcessedQueryList.MemoryCache)
```

Comment:
Its not immediately clear why DomainIterator and IntentIterator have the special __init__ function but `EntitiesIterator ` does ot

Reply:
It's because domain and intent are small bits of data that are cached in memory.  When they are run through multiple times, the data is used from memory instead of reading from the db again.